### PR TITLE
Feat: to_csv

### PR DESCRIPTION
**Pull request by kazantsev-maksim**
## Which issue does this PR close?

- N/A

## Rationale for this change

Basic implementation of the spark to_csv function added - https://spark.apache.org/docs/latest/api/sql/index.html#to_csv

1. Handling of complex types must be implemented in a future iteration.
2. The processing of types such as DateType, TimestampType, TimestampNTZType, and BinaryType is currently inconsistent with Spark's behavior.

## How are these changes tested?

1. Added unit tests
3. Added benchmark tests

Benchmark results (need optimization):
to_csv_benchmark_result

### CODE REVIEW

File: native/proto/src/proto/expr.proto [hunk 1]
```diff
@@ -275,6 +276,22 @@ message FromJson {
   string timezone = 3;
 }
 
+message ToCsv {
+  Expr child = 1;
+  CsvWriteOptions options = 2;
+}
+
+message CsvWriteOptions {
```
Comment (kazantsev-maksim) [inline native/proto/src/proto/expr.proto:285]: All settings: https://spark.apache.org/docs/latest/sql-data-sources-csv.html

File: native/spark-expr/src/csv_funcs/to_csv.rs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        }
+    }
+    Ok(Arc::new(builder.finish()))
+}
+
+#[inline]
+fn escape_value(value: &str, quote: &str, escape: &str, output: &mut String) {
+    for ch in value.chars() {
+        let ch_str = ch.to_string();
+        if ch_str == quote || ch_str == escape {
```
Comment (parthchandra) [inline native/spark-expr/src/csv_funcs/to_csv.rs]: The CSV spec does not have any special escape for escape and the preferred way to escape the double quote is another double quote (but only if the string is enclosed in a double quote!). - https://datatracker.ietf.org/doc/html/rfc4180#section-2
Not sure what Spark does here.

**Review by parthchandra — APPROVED** [review]
(Sorry for the delay in reviewing). This looks pretty good to me, pending ci.
Also a minor comment on escaping. Can you confirm that this behaviour is consistent with Spark?

File: native/spark-expr/src/csv_funcs/to_csv.rs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                        value = value.trim_start();
+                    }
+                    if write_options.ignore_trailing_white_space {
+                        value = value.trim_end();
+                    }
+                }
+                let needs_quoting = write_options.quote_all
+                    || (is_string_field
+                        && !string_arrays[col_idx].is_null(row_idx)
+                        && (value.contains(&write_options.delimiter)
```
Comment (parthchandra) [inline native/spark-expr/src/csv_funcs/to_csv.rs]: If not ignoring leading/trailing whitespace then the string should be quoted if there is a leading/trailing whitespace. 
Also if there are newlines.

**Comment by kazantsev-maksim** [comment]
Tanks @parthchandra for the review. I'll try to resolve conversions them tomorrow.

File: native/spark-expr/src/csv_funcs/to_csv.rs (hunk 3 above)
Comment (kazantsev-maksim) [reply native/spark-expr/src/csv_funcs/to_csv.rs]: Spark don't quoting string with leading or trailing whitespaces if quoteAll option is disabled. This edge case is covered by unit tests

```sql 
sql(s"insert` into $table values('  abc   ')")
```

```sql
checkSparkAnswerAndOperator(
          df.select(
            to_csv(
              struct(col("col"), lit(1)),
              Map(
                "delimiter" -> ";",
                "ignoreLeadingWhiteSpace" -> "false",
                "ignoreTrailingWhiteSpace" -> "false").asJava)))
 ```

Comment (kazantsev-maksim) [reply native/spark-expr/src/csv_funcs/to_csv.rs]: For case of newlines you are absolutley right - fixed this.

File: native/spark-expr/src/csv_funcs/to_csv.rs (hunk 2 above)
Comment (kazantsev-maksim) [reply native/spark-expr/src/csv_funcs/to_csv.rs]: This necessary for case: 

```sql
sql(s"""insert into $table values('a\\\\b')""")
```

```sql
checkSparkAnswerAndOperator(
          df.select(to_csv(struct(col("col"), lit(1)), Map("quoteAll" -> "true").asJava)))
```

This may seem insignificant, but I want to cover the maximum number of edge cases.

**Comment by kazantsev-maksim** [comment]
@parthchandra could you take another look when you have time?

File: native/spark-expr/src/csv_funcs/csv_write_options.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    pub escape: String,
+    pub null_value: String,
+    pub quote_all: bool,
+    pub ignore_leading_white_space: bool,
+    pub ignore_trailing_white_space: bool,
+}
+
+impl Display for CsvWriteOptions {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
```
Comment (andygrove) [inline native/spark-expr/src/csv_funcs/csv_write_options.rs:33]: nit: should `delimiter` also be shown here?

File: spark/src/main/scala/org/apache/comet/serde/structs.scala [hunk 5]
```diff
@@ -230,3 +232,68 @@ object CometJsonToStructs extends CometExpressionSerde[JsonToStructs] {
...
+          s"The schema ${expr.inputSchema} is not supported because it includes a complex type"))
+    }
+    val containsIncompatibleDataTypes = dataTypes.exists(incompatibleDataTypes.contains)
+    if (containsIncompatibleDataTypes) {
+      return Incompatible(
+        Some(
+          s"The schema ${expr.inputSchema} is not supported because " +
+            s"it includes a incompatible data types: $incompatibleDataTypes"))
+    }
+    Compatible()
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/serde/structs.scala]: Can we guarantee full compatibility with the `UnivocityGenerator` that Spark uses? Perhaps this should be marked as incompatible for now until we have sufficient fuzz testing to confirm compatibility?

File: native/spark-expr/src/csv_funcs/to_csv.rs [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        .fields()
+        .iter()
+        .map(|f| matches!(f.data_type(), DataType::Utf8 | DataType::LargeUtf8))
+        .collect();
+
+    let mut builder = StringBuilder::with_capacity(array.len(), array.len() * 16);
+    let mut csv_string = String::with_capacity(array.len() * 16);
+
+    let quote_char = write_options.quote.chars().next().unwrap_or('"');
+    let escape_char = write_options.escape.chars().next().unwrap_or('\\');
```
Comment (andygrove) [inline native/spark-expr/src/csv_funcs/to_csv.rs:154]: Does Spark limit these to chars, or can they be multi-character strings?

**Comment by andygrove** [comment]
@kazantsev-maksim This looks great, thank you! My only concern is that we are claiming full compatibility with Spark and I'm not sure that the tests are comprehensive enough to prove that. I am going to do some testing with this PR and see if I can suggest more tests to add.

**Comment by kazantsev-maksim** [comment]
Thanks @andygrove @parthchandra for the review. I will continue working on this function to achieve full compatibility with Spark.

File: native/spark-expr/src/csv_funcs/csv_write_options.rs (hunk 4 above)
Comment (kazantsev-maksim) [reply native/spark-expr/src/csv_funcs/csv_write_options.rs:33]: Thanks, fixed.

File: spark/src/main/scala/org/apache/comet/serde/structs.scala (hunk 5 above)
Comment (kazantsev-maksim) [reply spark/src/main/scala/org/apache/comet/serde/structs.scala]: Thanks, fixed.

**Comment by andygrove** [comment]
@kazantsev-maksim oops, I was going through the PR queue and saw that this one had two approvals, and I went ahead and merged before noticing that you had continued working on this. It looks like the expression is already marked as incompatible, so this shouldn't be an issue.

**Comment by kazantsev-maksim** [comment]
Thanks @andygrove. I think I'll be able to continue working on the function as part of the issue: https://github.com/apache/datafusion-comet/issues/3232
